core/services/health/nurse.go

```diff
@@ -0,0 +1,382 @@
+	return n.StartOnce("nurse", func() error {
+		// This must be set *once*, and it must occur as early as possible
+		runtime.MemProfileRate = n.cfg.MemProfileRate
+
+		n.AddCheck("mem", n.checkMem)
+		n.AddCheck("goroutines", n.checkGoroutines)
+
+		n.wgDone.Add(2)
+
+		// Checker
+		go func() {
+			defer n.wgDone.Done()
+			for {
+				select {
+				case <-n.chStop:
+					return
+				case <-time.After(n.cfg.PollInterval):
+				}
+
+				func() {
+					n.checksMu.RLock()
+					defer n.checksMu.RUnlock()
+					for reason, checkFunc := range n.checks {
+						if unwell, meta := checkFunc(); unwell {
+							n.GatherVitals(reason, meta)
+							break
+						}
+					}
+				}()
+			}
+		}()
+
+		// Responder
+		go func() {
+			defer n.wgDone.Done()
+			for {
+				select {
+				case <-n.chStop:
+					return
+				case req := <-n.chGather:
+					n.gatherVitals(req.reason, req.meta)
+				}
+			}
+		}()
+
+		return nil
+	})
+}
+
+func (n *Nurse) Close() error {
+	return n.StopOnce("nurse", func() error {
+		close(n.chStop)
+		n.wgDone.Wait()
+		return nil
+	})
+}
+
+func (n *Nurse) AddCheck(reason string, checkFunc CheckFunc) {
+	n.checksMu.Lock()
+	defer n.checksMu.Unlock()
+	n.checks[reason] = checkFunc
+}
+
+func (n *Nurse) GatherVitals(reason string, meta Meta) {
+	select {
+	case <-n.chStop:
+	case n.chGather <- gatherRequest{reason, meta}:
+	default:
+	}
+}
+
+func (n *Nurse) checkMem() (bool, Meta) {
+	var memStats runtime.MemStats
+	runtime.ReadMemStats(&memStats)
+	unwell := memStats.Alloc >= uint64(n.cfg.MemThreshold)
+	if !unwell {
+		return false, nil
+	}
+	return true, Meta{
+		"mem_alloc": utils.FileSize(memStats.Alloc),
+		"threshold": n.cfg.MemThreshold,
+	}
+}
+
+func (n *Nurse) checkGoroutines() (bool, Meta) {
+	num := runtime.NumGoroutine()
+	unwell := num >= n.cfg.GoroutineThreshold
+	if !unwell {
+		return false, nil
+	}
+	return true, Meta{
+		"num_goroutine": num,
+		"threshold":     n.cfg.GoroutineThreshold,
+	}
+}
+
+func (n *Nurse) gatherVitals(reason string, meta Meta) {
+	loggerFields := (logger.Fields{"reason": reason}).Merge(logger.Fields(meta))
+
+	n.log.Debugw("Nurse is gathering vitals", loggerFields.Slice()...)
+
+	size, err := n.totalProfileBytes()
+	if err != nil {
+		n.log.Errorw("could not fetch total profile bytes", loggerFields.With("error", err).Slice()...)
+		return
+	} else if size >= uint64(n.cfg.MaxProfileSize) {
+		n.log.Warnw("cannot write pprof profile, total profile size exceeds configured PPROF_MAX_PROFILE_SIZE",
+			loggerFields.With("total", size, "max", n.cfg.MaxProfileSize).Slice()...,
+		)
+		return
+	}
+
+	runtime.SetCPUProfileRate(n.cfg.CPUProfileRate)
+	defer runtime.SetCPUProfileRate(0)
+	runtime.SetBlockProfileRate(n.cfg.BlockProfileRate)
+	defer runtime.SetBlockProfileRate(0)
+	runtime.SetMutexProfileFraction(n.cfg.MutexProfileFraction)
+	defer runtime.SetMutexProfileFraction(0)
+
+	now := time.Now()
+
+	var wg sync.WaitGroup
+	wg.Add(9)
+
+	go n.appendLog(now, reason, meta, &wg)
```

Comment:
It might be good to do `appendLog` synchronously on this goroutine, to ensure that `nurse.log` is updated before anything else.

And I suppose if that actually errors writing to the filesystem, then it could make sense to short circuit everything else too.